### ISSUE TITLE
Cross-validation code

### ISSUE THREAD

**Issue by 14corman** [14corman/AIDescisionTree#11]
@the5thbeatle I just had a look at your cross validation code, and there are a few problems that will come up when you try to run it:

1. You should not make Feature_map. You need to use the 1 provided in DTree.
2. You should not open TreeClassifier. This code will be opened and used in TreeClassifier. TreeClassifier is the open point to everything.
3. I think I may have a more simpler shuffle method that I worked on yesterday that I can commit now so that you can see how I did it. I am not saying yours is not right, but just by looking at it I cannot see any guarantee that each shuffle for each feature will be the same. This is because of the Random boolean you are using to partition a front and back of each list. 

When you look at my code, feel free to delete any and everything that you will not need or use. My code also shows how TreeClassifier should be used. It is incomplete, so do not try to run it. It will show that there are non complete sections to it.

EDIT:
This code is in the metrics file that I just uploaded.

**Comment by the5thbeatle** [comment]
That code was unfinished as of yet. I know I won’t need to make a feature
map module for the real deal, but like I said, I can’t get any files with
dependencies tested, so I had that definition as a stand in.

I found your message about the shuffler after I started that file. That’s
why I stopped with that function and quit working for the night. I’ll look
into switching it over to yours today.

You probably also saw that there’s an OCaml csv parser now, so that can be
used as well.

**Comment by 14corman** [comment]
How is the code for cross-validation going? Do you need any help with any of it?

**Comment by the5thbeatle** [comment]
I had to spend a fair amount of time on the function that shuffles the
feature map, but I got it working. All of the lists of values are shuffled
in the same was as the list of class labels. I'm revising the parser code a
small amount, and then I'll push what I have. What's leftover is tough to
do until I figure out how to test it with the actual tree code.

**Comment by 14corman** [comment]
Ok sounds good. I will wait for your push and then convert the project to dune. With dune we should be able to test code in the visual studio terminal which will be a life saver.

Cory Kromer-Edwards, B.A
University of Iowa
Computer Science PhD Student
Graduate Teaching Assistant-Data Structures


On Apr 30, 2019, at 12:22 PM, the5thbeatle <notifications@github.com<mailto:notifications@github.com>> wrote:

I had to spend a fair amount of time on the function that shuffles the
feature map, but I got it working. All of the lists of values are shuffled
in the same was as the list of class labels. I'm revising the parser code a
small amount, and then I'll push what I have. What's leftover is tough to
do until I figure out how to test it with the actual tree code.

**Comment by the5thbeatle** [comment]
I've pushed the code. It's still in a separate branch, but I can pull it
into master if you'd like.

**Comment by 14corman** [comment]
Sure please. Merge your branch to master. I will then convert over to dune. Then you can continue where you left off.

Cory Kromer-Edwards, B.A
University of Iowa
Computer Science PhD Student
Graduate Teaching Assistant-Data Structures


On Apr 30, 2019, at 12:33 PM, the5thbeatle <notifications@github.com<mailto:notifications@github.com>> wrote:

I've pushed the code. It's still in a separate branch, but I can pull it
into master if you'd like.

**Comment by the5thbeatle** [comment]
Just merged it.

**Comment by 14corman** [comment]
Is the cross validation code complete? If so what does it return? We may talk about this in the meeting, but I wanted to post this in case I forget to ask.

**Comment by the5thbeatle** [comment]
It isn’t yet. I’ve gotten tied up in deciding what to return and when to
compute things. I’ve got matrices of predicted vs actual label counts. From
that, overall specificity and sensitivity can be calculated. After all is
said and done, it will obviously print a graph. (Which reminds me — I don’t
have the PDF open currently, but iirc, one of the functions should accept a
graph file as an input). Should the cross validation really do much besides
printing the results, printing the tree, and maybe saving the tree?

**Comment by the5thbeatle** [comment]
Side note. B11 is being used for office hours, it looks like.

**Comment by 14corman** [comment]
We are in lounge.

**Comment by the5thbeatle** [comment]
Can you point me in the right direction?

**Comment by 14corman** [comment]
Its by 14 (main offices) in MLH.

**Comment by 14corman** [comment]
Have you had any luck with the new method to cross validation?

**Comment by 14corman** [comment]
Also, Jace and I will be in the lounge at 11 for a little while if you want to/need to stop by to explain anything or need help with anything.

**Comment by the5thbeatle** [comment]
I've taken the cross validation work into a separate function. Now
cross_validate and the model selector both call that separate function. The
thing is that it now needs to accumulate some training error to fit with
the book's algorithm, so that separate function is going to output more
data. I said that cross_validate wouldn't change in its inputs and outputs,
but to do that, I would be throwing away that training error.

**Comment by 14corman** [comment]
Since you do not have training error, we said you would want to maximize the F1 score instead. Are you doing an entire run of cross validation per depth, or are you doing a single fold per depth? If it is an entire run of cross validation then you just need 1 function to call cross validation with from depths 0 to max_depth and pick the depth with the highest F1 score.

**Comment by 14corman** [comment]
or until 2 graphs equal

**Comment by the5thbeatle** [comment]
I believe the project PDF explicitly mentions training error. It should be
one or two lines extra during the cross validation process and one float
being returned with the results. The depth selection process is restricted
to using depth=1 to 10 per the PDF as well. I’ll push something by the end
of the day that I think matches the project description and the book’s
depth selection.

**Comment by 14corman** [comment]
if you want to use error than just take 1 minus the accuracy since that is what it is.

**Comment by the5thbeatle** [comment]
I haven't done the depth selection yet (I found a problem with my previous
plan), but I've pushed the changes to the current cross validation. Cross
validate takes the same parameters but returns an extra float that
represents the average training error.

**Comment by 14corman** [comment]
You should use the accuracy measure from Metrics to perform accuracy then take 1 - accuracy to get error. Do you mean to have the model selector chose using single folds rather than a cross validation run? If so then you are ok, but you should not return average error rate the way you do.

If you do mean to have the model selector cross validation run do the model selector then you should do something like this:
``` ocaml
let selector (k : int) (d : int option) (feature_map,labels,num_feature_values) = 
    let rec temp_selector (k : int) (d : int option) (feature_map,labels,num_feature_values) = 
        let (g, predicted, actual) = cross_validator k d (feature_map,labels,num_feature_values) in
        let error = 1.0 - accuracy predicted actual [attr] in
        let (g2, predicted2, actual2, error2) = temp_selector k (d - 1) (feature_map,labels,num_feature_values) in
        if error < error2 then
             (g, predicted, actual, error)
        else
             (g2, predicted2, actual2, error2)
    in temp_selector k d (feature_map,labels,num_feature_values)
```

**Comment by the5thbeatle** [comment]
The problem isn't calculating the validation error, it's the training
error. The cross validation process didn't count how many examples appeared
with the wrong label in a leaf node, so that had to be added.

A new problem, though, is that setting the depth of the tree equal to 0 or
1 gives a training error that's as good as a fully split tree. For example,
with the house votes dataset, depth=Some 0 or depth=Some 1 gives a training
error of 0.0000, which shouldn't be possible. I'm also having trouble with
when to decide that the training error has converged. My first thought was
to just run the validator from depth=0 to depth=(# features), but that's
not really keeping with the spirit of the book's algorithm.

**Comment by the5thbeatle** [comment]
Looking at the accuracy definition, why does it have (tp + fp) / (#
examples)? Shouldn't it be (tp + tn)?

**Comment by the5thbeatle** [comment]
Okay, a couple of things. First, the with depth=(Some 0) and depth=(Some
1), the tree is built until it can't split anymore. I would expect
depth=(Some 0) to essentially guess the majority class label and
depth=(Some 1) to have only one split.

The other thing is that I don't know of a good heuristic to check for the
training error to converge.

Part of the assignment is to run the model selector on two datasets with
parameters from depth=1 to depth=10. We then have to plot the training
errors and the validation errors on a graph and explain which depth was
chose for each dataset. That is possible at this point, since the bounds
for the depth are given. However, the model selector doesn't function if
the max depth isn't input. I'm not exactly sure what to do about it, so I'm
pushing what I have until I have a new idea. If either of you comes up with
a solution, take a whack at it.

There is a main.ml file that runs the model selector from depth=1 to
depth=10.

**Comment by 14corman** [comment]
Since I fixed the depth bug, I thought I would build the graph of training and validation error so everyone can see for the house votes dataset. As you can see, it started to overfit pretty badly after depth 1.

![image](https://user-images.githubusercontent.com/15676385/57170907-a0845700-6dd5-11e9-8f0d-c0fbc0b625eb.png)

Why not build models from 1 to max depth (which will be 10) and then take the depth with the lowest validation error?

**Comment by the5thbeatle** [comment]
The generalized process is to build the tree until it stops improving with
training error. After that, it’s essentially guaranteed that you’re going
to overfit. After you find that threshold of certain overfitting, that’s
when you look for the depth that produced the best validation error.

The specialized case is when we input the desired depth range. However, we
still can’t just scan the output for the lowest validation error, since
that could mean taking a tree from beyond the overfitting threshold. It
being the best tree that the model selector found would be an anomaly,
which isn’t ideal for selection. We would still need to find that threshold
to maintain correctness.

**Comment by 14corman** [comment]
Could you use a threshold to stop if validation error - training error goes above this? Say for the graph above, if we have the threshold to be about 0.00001 then we would have stopped at depth 3. I am working on the code now, and could put that in if we think that would work.

**Comment by 14corman** [comment]
I set up the threshold like this: 
if val_error -. t_err > 0.005
Then, I tried with the house votes dataset from 1 to 10, and it did 1 and 2 before stopping like I said it would in the above comment.
```
training errs: [0.030172; 0.030172; ]
validate errs: [(1, 0.030172); (2, 0.030172); ]
```

depth 1 won, and here is the tree that was returned:
``` dot
digraph G {
  1 [label=<physician-fee-freeze<br/>remainder: 0.181745319882<br/>Samples:[124,108,]>,
     shape=box, style="rounded,filled", ];
  2 [label=<1<br/>Samples:[6,107,]>, shape=box, style="rounded,filled", ];
  3 [label=<0<br/>Samples:[118,1,]>, shape=box, style="rounded,filled", ];


  1 -> 2 [label="1", ];
  1 -> 3 [label="0", ];

  }
```

![graphviz (13)](https://user-images.githubusercontent.com/15676385/57172367-7be3ab80-6de4-11e9-86f2-676b65045952.png)

The test F1 score is 0.96.

**Comment by 14corman** [comment]
I will push the code now so you can see it.

**Comment by the5thbeatle** [comment]
That might be the best we have. I’ve never had to do this part manually, so
I don’t know if there’s some function of the last n training errors that
could work. That might be a good enough delta to use.

**Comment by 14corman** [comment]
Well we would want to do something like mean squared error, but the problem with this is that the number of iterations is incredibly small (between 1 and 10). With something like a neural network you may have hundreds if not thousands of iterations. I think that is why it is so hard to find a good convergence measure. I just pushed the code, so see what you think.

**Comment by 14corman** [comment]
I think we should keep the list of validation and test results print out so we can graph that for the paper.

**Comment by the5thbeatle** [comment]
There are still two things that I think are problematic with the solution.
The first one is somewhat trivial. When a max depth isn't supplied to the
model selector, instead of 10 being the upper limit, the number of features
should be, as that's the max depth of the tree.

The other thing, though, is that the check for training error convergence
shouldn't be a function of the validation error. It should only look at the
current training error and the past training errors. I've pushed a change
that checks the previous training error against the current one; if the
difference is less than what we'd see with only two examples changing, then
it stops going deeper. I also added a check for the current depth being
equal to the max depth. There's a possibility that instead of two examples,
we could select a certain percentage of examples being the threshold.

**Comment by 14corman** [comment]
There are 2 problems with checking previous to current training error which is why I did not do it:

1. The training errors are already incredibly small from the start, so the threshold may always be hit at depth 1 or 2.

2. The difference between 2 training errors could be less than 0.00001 which at that point it would be hard to check for, and I actually tried this and it never passed no matter how small I made the threshold.

Another thing is you do not want to check the absolute value of the difference. If the error starts going up then checking the absolute value will have it keep going until the end. You want to make sure the current error is so much less than the previous one.

Where did the 8 and 3 come from for epsilon?

I thought we wanted to see if it was starting to overfit. If you look at the chart from above you will see that if the threshold does not hit from depth 1 to 2 then it will train all the way to 10.

The method needs to end in a call to cross validator so that the method will return the proper stuff for the winning depth. If you want to test this further use the BuildDTreeDepth file as that has this implemented. I will delete main since that is unneeded now.

Good catch with the check for if d is at max depth. I forgot to add that.

The code should fail if a max depth is not given. You should only worry about a min depth. The BuildDTreeDepth file throws an error if -d if not in the input arguments for max depth. If you want to put that in selector method than ok, but this would be best I think.

Lastly, do we want to include the error values when we stop the recursive calls? At that point it would start to overfit, so it should not be added I think.

**Comment by the5thbeatle** [comment]
The 8 and the 3 from epsilon com from       2 examples / (.75 = ( (k-1)/k ))
It's just a small bit of algebra to clean it up a tad bit.

The problem isn't that we're checking current training error against a past
one(s). That's part of the algorithm we're supposed to be implementing. The
problem is we just don't know of a good way to reliably check for
convergence.

Incredibly small training errors in the beginning of the search only
terminate the search early if the training errors won't get much better
anyway. I don't see a problem with that.

Checking for convergence generally does follow the "set an epsilon, does my
current value differ from my previous from less than epsilon" pattern, and
I think we should adhere to that. I will admit that 2 examples is rather
small, though, especially for larger datasets. That's why I think we should
consider using a percentage of the training examples instead of the
arbitrary count that I chose. A percentage is at least somewhat relativized
for each dataset.

I don't think I see what you mean for not checking the absolute value. The
training error should never go up significantly, so that training ad
infinitum shouldn't happen. Checking the absolute value, though, should
account for any random but insignificant rise in training error that occurs
between two equivalent models.

**Comment by the5thbeatle** [comment]
Also, the code shouldn't fail on lack of max depth. The book's algorithm
goes unbounded, so that's what I attempted to do as well.

I figured we could include the final error values. It's just one set of
values, and we don't know exactly what the error curves look like a priori,
so that last value could be better than the previous one in validation
error. If it is, then yay. If it isn't, then I don't see any harm.

**Comment by the5thbeatle** [comment]
However, I did do the upper bound wrong for depth. I meant to do the number
of features, not the number of examples.

**Comment by 14corman** [comment]
Ok. I will fix BuildDTreeDepth to not fail if a max depth is not given. I see where you are coming from for the epsilon. However, when you say 2 examples you do not mean examples from the data, but instead 2 training error examples right? You say you use 2 examples, which makes sense if you talk about this and the previous training error, but then you say you should use a percentage of examples for epsilon. Do you still mean a percentage of training errors to check?

I just fixed the depth checking for number of features instead of number of examples from the labels.

I think you are calculating the training error if that is what you mean by needing to take a percentage.

**Comment by 14corman** [comment]
I just made the changes and ran with the changes you made. Here are the results:
```
Working on dataset: house-votes-84
Epsilon: 0.011494
training errs: [0.030172; 0.030172; ]
validate errs: [(1, 0.030172); (2, 0.030172); ]
winning depth: 1
digraph G {
  1 [label=<physician-fee-freeze<br/>remainder: 0.181745319882<br/>Samples:[124,108,]>,
     shape=box, style="rounded,filled", ];
  2 [label=<1<br/>Samples:[6,107,]>, shape=box, style="rounded,filled", ];
  3 [label=<0<br/>Samples:[118,1,]>, shape=box, style="rounded,filled", ];


  1 -> 2 [label="1", ];
  1 -> 3 [label="0", ];

  }
```

It seemed to stop at depth = 2 which is to be expected.

**Comment by 14corman** [comment]
I ran that with max depth of 10

**Comment by the5thbeatle** [comment]
I think there may have been some miscommunication on my part. When I said
the program shouldn't fail with max depth not supplied, I meant that the
model selector shouldn't fail. If the depth-supplied tree builder fails
without a depth input, that is likely okay.

Yes, I had it so that the two example difference was for two training
errors. And yes, I think we could consider using a percentage difference in
training errors between depths instead of a predetermined example count.
For a dataset with 100 examples, that 2 example difference might be a good
indicator of convergence, but for a dataset with 1000 examples, we would
probably be well beyond the threshold of overfitting.

I think that it might be good to set epsilon to 1% or 0.5% of the number of
training examples.

There's one more thing I can think to fix. If the user enters a max depth
for the depth selector, do we ignore the training error convergence? It
would make sense to if we are expected to graph ten pairs of errors for two
datasets.

**Comment by 14corman** [comment]
Ok I understand what you are saying now. I think the convergence part is good to implement in a library setting, but as you say we are expected to make a file give a max depth which you would expect the program to go all the way through.

There is not much out there for convergence on decision trees. This is all I could find, and I do not think we have enough time to go through and read and understand all of it to implement (let alone from scratch): [http://papers.nips.cc/paper/2625-on-the-adaptive-properties-of-decision-trees.pdf](http://papers.nips.cc/paper/2625-on-the-adaptive-properties-of-decision-trees.pdf).

We could just say in the notes that even though the book wanted us to make a method to converge, we thought that since a max depth had to be given that we thought it as the user wanting to go over all depth options rather than starting early. Would that work?

**Comment by 14corman** [comment]
Or we could have 2 selector functions. 1 that converges and 1 that goes to a max depth. Then we could use the max depth option for the problem and say in the notes that we had 1 to converge as the book says.

**Comment by the5thbeatle** [comment]
How about this. We can set a default convergence percentage to determine
epsilon, but allow a percentage to be input instead of our default. That
gives flexibility on that route. In addition, if a user inputs a max depth,
no convergence is checked at all, meaning we can get all ten measurements
for the graph. That gives full control to the user while still allowing us
to satisfy requirements.

**Comment by 14corman** [comment]
Sounds like a plan to me.

Cory Kromer-Edwards, B.A
University of Iowa
Computer Science PhD Student
Graduate Teaching Assistant-Data Structures


On May 4, 2019, at 1:02 PM, the5thbeatle <notifications@github.com<mailto:notifications@github.com>> wrote:

How about this. We can set a default convergence percentage to determine
epsilon, but allow a percentage to be input instead of our default. That
gives flexibility on that route. In addition, if a user inputs a max depth,
no convergence is checked at all, meaning we can get all ten measurements
for the graph. That gives full control to the user while still allowing us
to satisfy requirements.

**Comment by the5thbeatle** [comment]
I think I fixed the last couple of things. Look it over and let me know if
there's something I missed or if there's something we're not meeting in the
project specification.

**Comment by 14corman** [comment]
It looks and works great. I think with that we should be done coding everything.

**Comment by the5thbeatle** [comment]
Do we still need some programs to run them? A DTL.ml to build a tree and a
ModelSelection.ml to pick a depth? Just two files for executables that can
take the command line args for them to test with and grade?

**Comment by 14corman** [comment]
We have had all of those files, and talked about this, for a while now. They are named:
* BuildDTree.ml
* BuildDTreeDepth.ml
* Classifier.ml

**Comment by the5thbeatle** [comment]
My mistake. I missed those.
